File: Code/Avernakis Nodejs/Test-Nodejs/src/Ave/Ui/Control/UiMatrixEditor.ts
import { IControl } from "../UiControl";
import { AveLib } from "../../AveLib";
import { Window } from "./UiWindow";
import {
    DpiSize,
    DpiSize_2,
    KnobControlType,
    Rect,
    IconCache,
} from "../UiCommon";
import { Vec4, Vec2 } from "../../Math/Vector";

export enum MatrixUnitType {
    Empty /**/,
    Switch /**/,
    Pointer /**/,
    Value /**/,
    Pie /**/,
    Pan /**/,
    Icon /**/,
    Color /**/,
}

export enum MatrixUnitFlag {
    Default /**/ = 0x0,
    Disabled /**/ = 0x1,
    Hidden /**/ = 0x2,
    Active /**/ = 0x4,
}

export class MatrixUnit {
    Type: MatrixUnitType = MatrixUnitType.Empty;
    Flag: MatrixUnitFlag = MatrixUnitFlag.Default;
    DataValue: number = 0;
    DataIcon: IconCache = 0;
    DataColor: Vec4 = new Vec4();
    BackColor: Vec4 = new Vec4();
    ForeColor: Vec4 = new Vec4();
    FillColor: Vec4 = new Vec4();
}

export interface IMatrixEditorDoc {
    SetDimension(vSize: Vec2): IMatrixEditorDoc;
    GetDimension(): Vec2;

    SetVirtualUnit(
        fn: (sender: IMatrixEditorDoc, x: number, y: number) => MatrixUnit
    ): IMatrixEditorDoc;
    SetVirtualColHeader(
        fn: (sender: IMatrixEditorDoc, n: number) => string
    ): IMatrixEditorDoc;
    SetVirtualRowHeader(
        fn: (sender: IMatrixEditorDoc, n: number) => string
    ): IMatrixEditorDoc;

    SetUnit(x: number, y: number, u: MatrixUnit): IMatrixEditorDoc;
    GetUnit(x: number, y: number): MatrixUnit;
    SetUnitBackColor(vColor: Vec4): IMatrixEditorDoc;
    GetUnitBackColor(): Vec4;
    SetUnitForeColor(vColor: Vec4): IMatrixEditorDoc;
    GetUnitForeColor(): Vec4;
    SetUnitFillColor(vColor: Vec4): IMatrixEditorDoc;
    GetUnitFillColor(): Vec4;

    ColClearName(): IMatrixEditorDoc;
    ColSetName(n: number, s: string): IMatrixEditorDoc;
    ColGetName(n: number): string;
    ColClearSplitter(): IMatrixEditorDoc;
    ColSetSplitter(n: number, nSize: DpiSize): IMatrixEditorDoc;
    ColGetSplitter(n: number): DpiSize;
    ColSetIcon(n: number, nIcon: IconCache): IMatrixEditorDoc;
    ColGetIcon(n: number): IconCache;
    ColSetBackColor(n: number, vColor: Vec4): IMatrixEditorDoc;
    ColGetBackColor(n: number): Vec4;
    ColSetVisible(n: number, b: boolean): IMatrixEditorDoc;
    ColGetVisible(n: number): boolean;
    ColSetEnable(n: number, b: boolean): IMatrixEditorDoc;
    ColGetEnable(n: number): boolean;

    RowClearName(): IMatrixEditorDoc;
    RowSetName(n: number, s: string): IMatrixEditorDoc;
    RowGetName(n: number): string;
    RowClearSplitter(): IMatrixEditorDoc;
    RowSetSplitter(n: number, nSize: DpiSize): IMatrixEditorDoc;
    RowGetSplitter(n: number): DpiSize;
    RowSetIcon(n: number, nIcon: IconCache): IMatrixEditorDoc;
    RowGetIcon(n: number): IconCache;
    RowSetBackColor(n: number, vColor: Vec4): IMatrixEditorDoc;
    RowGetBackColor(n: number): Vec4;
    RowSetVisible(n: number, b: boolean): IMatrixEditorDoc;
    RowGetVisible(n: number): boolean;
    RowSetEnable(n: number, b: boolean): IMatrixEditorDoc;
    RowGetEnable(n: number): boolean;

    ViewSetValueFormat(
        fn: (
            sender: IMatrixEditorDoc,
            x: number,
            y: number,
            v: number
        ) => string
    );

    Redraw(): IMatrixEditorDoc;
}

export interface IMatrixEditor extends IControl {
    new (window: Window): IMatrixEditor;

    DocNew(): MatrixEditor;
    DocGet(): IMatrixEditorDoc;
    DocLink(p: MatrixEditor): MatrixEditor;

    ViewReset(): MatrixEditor;
    ViewCopyFrom(p: MatrixEditor): MatrixEditor;

    ColSetSizeAll(nSize: DpiSize): MatrixEditor;
    ColGetSizeAll(): DpiSize;
    ColSetSize(n: number, nSize: DpiSize): MatrixEditor;
    ColGetSize(n: number): DpiSize;
    ColSetVertical(b: boolean): MatrixEditor;
    ColGetVertical(): boolean;

    RowSetSizeAll(nSize: DpiSize): MatrixEditor;
    RowGetSizeAll(): DpiSize;
    RowSetSize(n: number, nSize: DpiSize): MatrixEditor;
    RowGetSize(n: number): DpiSize;

    ViewSetCornerSize(vSize: DpiSize_2): MatrixEditor;
    ViewGetCornerSize(): DpiSize_2;
    ViewSetCornerContent(c: IControl): IControl;
    ViewGetCornerContent(): IControl;
    ViewSetBack(b: boolean): MatrixEditor;
    ViewGetBack(): boolean;
    ViewSetUnit(b: boolean): MatrixEditor;
    ViewGetUnit(): boolean;
    ViewSetPanel(b: boolean): MatrixEditor;
    ViewGetPanel(): boolean;
    ViewSetColHeader(b: boolean): MatrixEditor;
    ViewGetColHeader(): boolean;
    ViewSetRowHeader(b: boolean): MatrixEditor;
    ViewGetRowHeader(): boolean;
    ViewSetUnitBack(b: boolean): MatrixEditor;
    ViewGetUnitBack(): boolean;
    ViewSetCursor(b: boolean): MatrixEditor;
    ViewGetCursor(): boolean;
    ViewSetCursorEmpty(b: boolean): MatrixEditor;
    ViewGetCursorEmpty(): boolean;
    ViewSetGridX(b: boolean): MatrixEditor;
    ViewGetGridX(): boolean;
    ViewSetGridY(b: boolean): MatrixEditor;
    ViewGetGridY(): boolean;
    ViewSetScrollX(b: boolean): MatrixEditor;
    ViewGetScrollX(): boolean;
    ViewSetScrollY(b: boolean): MatrixEditor;
    ViewGetScrollY(): boolean;

    UnitGetRect(x: number, y: number): Rect;

    SetKnobControlType(nType: KnobControlType): MatrixEditor;
    GetKnobControlType(): KnobControlType;

    GetEditorRect(): Rect;

    OnUnitClick(
        fn: (sender: MatrixEditor, x: number, y: number, u: MatrixUnit) => void
    ): MatrixEditor;
    OnUnitRightClick(
        fn: (sender: MatrixEditor, x: number, y: number, u: MatrixUnit) => void
    ): MatrixEditor;
    OnUnitChange(
        fn: (sender: MatrixEditor, x: number, y: number, u: MatrixUnit) => void
    ): MatrixEditor;
    OnUnitChangeBegin(
        fn: (sender: MatrixEditor, x: number, y: number, u: MatrixUnit) => void
    ): MatrixEditor;
    OnUnitChangeEnd(
        fn: (
            sender: MatrixEditor,
            x: number,
            y: number,
            u: MatrixUnit,
            bCanceled: boolean
        ) => boolean
    ): MatrixEditor; // return false to reject the change, bCanceled is the recommanded behavior.
}

export class MatrixEditor extends (AveLib.UiMatrixEditor as IMatrixEditor) {
    private m_CornerContent: IControl;

    DocGet(): IMatrixEditorDoc {
        const doc = super.DocGet();

        const OriginalGetDimension = doc.GetDimension.bind(doc);
        doc.GetDimension = () => Vec2.FromNative(OriginalGetDimension());

        const OriginalGetUnitBackColor = doc.GetUnitBackColor.bind(doc);
        doc.GetUnitBackColor = () =>
            Vec4.FromNative(OriginalGetUnitBackColor());

        const OriginalGetUnitForeColor = doc.GetUnitForeColor.bind(doc);
        doc.GetUnitForeColor = () =>
            Vec4.FromNative(OriginalGetUnitForeColor());

        const OriginalGetUnitFillColor = doc.GetUnitFillColor.bind(doc);
        doc.GetUnitFillColor = () =>
            Vec4.FromNative(OriginalGetUnitFillColor());

        const OriginalColGetBackColor = doc.ColGetBackColor.bind(doc);
        doc.ColGetBackColor = () => Vec4.FromNative(OriginalColGetBackColor());

        const OriginalRowGetBackColor = doc.RowGetBackColor.bind(doc);
        doc.RowGetBackColor = () => Vec4.FromNative(OriginalRowGetBackColor());

        return doc;
    }

    UnitGetRect(x: number, y: number): Rect {
        return Rect.FromNative(super.UnitGetRect(x, y));
    }

    GetEditorRect(): Rect {
        return Rect.FromNative(super.GetEditorRect());
    }

    ViewSetCornerContent(c: IControl) {
        this.m_CornerContent = c;
        return super.ViewSetCornerContent(c);
    }
}

File: Code/Avernakis Nodejs/Avernakis-Nodejs/Private/UiMatrixEditor.cpp
#include "StdAfx.h"
#include "UiMatrixEditor.h"
#include "UiMatrixEditorDoc.h"

#define ThisMethod($x) &UiMatrixEditor::$x
#define AutoAddMethod($x, ...) AddMethod<__VA_ARGS__>( #$x, ThisMethod( $x ) )
#define MakeThisFunc($x) MakeFunc( this, &UiMatrixEditor::$x )

namespace Nav
{

	namespace
	{
		ObjectRegister<UiMatrixEditor> c_obj;
	}

	void UiMatrixEditor::DefineControl()
	{
		//AutoAddMethod( SetText );


		AutoAddMethod( DocNew );
		AutoAddMethod( DocGet );
		AutoAddMethod( DocLink );

		AutoAddMethod( ViewReset );
		AutoAddMethod( ViewCopyFrom );

		AutoAddMethod( ColSetSizeAll );
		AutoAddMethod( ColGetSizeAll, WrapObjectGeneric );
		AutoAddMethod( ColSetSize );
		AutoAddMethod( ColGetSize, WrapObjectGeneric );
		AutoAddMethod( ColSetVertical );
		AutoAddMethod( ColGetVertical, WrapObjectGeneric );

		AutoAddMethod( RowSetSizeAll );
		AutoAddMethod( RowGetSizeAll, WrapObjectGeneric );
		AutoAddMethod( RowSetSize );
		AutoAddMethod( RowGetSize, WrapObjectGeneric );

		AutoAddMethod( ViewSetCornerSize );
		AutoAddMethod( ViewGetCornerSize, WrapObjectGeneric );
		AutoAddMethod( ViewSetCornerContent );
		AutoAddMethod( ViewGetCornerContent, WrapObjectGeneric );
		AutoAddMethod( ViewSetBack );
		AutoAddMethod( ViewGetBack, WrapObjectGeneric );
		AutoAddMethod( ViewSetUnit );
		AutoAddMethod( ViewGetUnit, WrapObjectGeneric );
		AutoAddMethod( ViewSetPanel );
		AutoAddMethod( ViewGetPanel, WrapObjectGeneric );
		AutoAddMethod( ViewSetColHeader );
		AutoAddMethod( ViewGetColHeader, WrapObjectGeneric );
		AutoAddMethod( ViewSetRowHeader );
		AutoAddMethod( ViewGetRowHeader, WrapObjectGeneric );
		AutoAddMethod( ViewSetUnitBack );
		AutoAddMethod( ViewGetUnitBack, WrapObjectGeneric );
		AutoAddMethod( ViewSetCursor );
		AutoAddMethod( ViewGetCursor, WrapObjectGeneric );
		AutoAddMethod( ViewSetCursorEmpty );
		AutoAddMethod( ViewGetCursorEmpty, WrapObjectGeneric );
		AutoAddMethod( ViewSetGridX );
		AutoAddMethod( ViewGetGridX, WrapObjectGeneric );
		AutoAddMethod( ViewSetGridY );
		AutoAddMethod( ViewGetGridY, WrapObjectGeneric );
		AutoAddMethod( ViewSetScrollX );
		AutoAddMethod( ViewGetScrollX, WrapObjectGeneric );
		AutoAddMethod( ViewSetScrollY );
		AutoAddMethod( ViewGetScrollY, WrapObjectGeneric );

		AutoAddMethod( UnitGetRect, WrapObjectGeneric );

		AutoAddMethod( SetKnobControlType );
		AutoAddMethod( GetKnobControlType, WrapObjectGeneric );

		AutoAddMethod( GetEditorRect, WrapObjectGeneric );

		AutoAddMethod( OnUnitClick, WrapObjectGeneric );
		AutoAddMethod( OnUnitRightClick, WrapObjectGeneric );
		AutoAddMethod( OnUnitChange, WrapObjectGeneric );
		AutoAddMethod( OnUnitChangeBegin, WrapObjectGeneric );
		AutoAddMethod( OnUnitChangeEnd, WrapObjectGeneric );
	}

	U1 UiMatrixEditor::Ctor( const CallbackInfo& ci, UiWindow * p, Napi::Value v )
	{
		m_Doc = ci.NewJsObjectWithOwnership<UiMatrixEditorDoc>();
		if ( !m_Doc )
			return false;

		if ( !__CreateControl( p, v ) )
			return false;

		m_Window = p;

		GetControlTyped().GetDoc().SetIconManager( p->PublicCloneIconManager() );
		m_Doc->SetDoc( &GetControlTyped().GetDoc() );

		return true;
	}

	void UiMatrixEditor::__OnUnitClick( Ui::IMatrixEditor & sender, S32 x, S32 y, const Ui::MatrixUnit & u )
	{
		m_OnUnitClick( this, x, y, u );
	}

	void UiMatrixEditor::__OnUnitRightClick( Ui::IMatrixEditor & sender, S32 x, S32 y, const Ui::MatrixUnit & u )
	{
		m_OnUnitRightClick( this, x, y, u );
	}

	void UiMatrixEditor::__OnUnitChange( Ui::IMatrixEditor & sender, S32 x, S32 y, const Ui::MatrixUnit & u )
	{
		m_OnUnitChange( this, x, y, u );
	}

	void UiMatrixEditor::__OnUnitChangeBegin( Ui::IMatrixEditor & sender, S32 x, S32 y, const Ui::MatrixUnit & u )
	{
		m_OnUnitChangeBegin( this, x, y, u );
	}

	void UiMatrixEditor::__OnUnitChangeEnd( Ui::IMatrixEditor & sender, S32 x, S32 y, const Ui::MatrixUnit & u, U1 & bCanceled )
	{
		U1 r = true;
		m_OnUnitChangeEnd.BlockCall( this, x, y, u, bCanceled, r );
		bCanceled = !r;
	}

	UiMatrixEditor * UiMatrixEditor::DocNew()
	{
		GetControlTyped().NewDoc();
		GetControlTyped().GetDoc().SetIconManager( m_Window->PublicCloneIconManager() );
		m_Doc->SetDoc( &GetControlTyped().GetDoc() );
		return this;
	}

	UiMatrixEditor * UiMatrixEditor::DocLink( UiMatrixEditor * p )
	{
		GetControlTyped().SetDoc( p->GetControlTyped().CloneDoc() );
		m_Doc->SetDoc( &GetControlTyped().GetDoc() );
		return this;
	}

	UiMatrixEditor* UiMatrixEditor::OnUnitClick       /**/( UnitCallback_t    /**/ && fn ) { m_OnUnitClick       /**/ = SetEventCallback<Ui::IMatrixEditor::OnUnitClick       /**/>( std::move( fn ), MakeThisFunc( __OnUnitClick       /**/ ) ); return this; }
	UiMatrixEditor* UiMatrixEditor::OnUnitRightClick  /**/( UnitCallback_t    /**/ && fn ) { m_OnUnitRightClick  /**/ = SetEventCallback<Ui::IMatrixEditor::OnUnitRightClick  /**/>( std::move( fn ), MakeThisFunc( __OnUnitRightClick  /**/ ) ); return this; }
	UiMatrixEditor* UiMatrixEditor::OnUnitChange      /**/( UnitCallback_t    /**/ && fn ) { m_OnUnitChange      /**/ = SetEventCallback<Ui::IMatrixEditor::OnUnitChange      /**/>( std::move( fn ), MakeThisFunc( __OnUnitChange      /**/ ) ); return this; }
	UiMatrixEditor* UiMatrixEditor::OnUnitChangeBegin /**/( UnitCallback_t    /**/ && fn ) { m_OnUnitChangeBegin /**/ = SetEventCallback<Ui::IMatrixEditor::OnUnitChangeBegin /**/>( std::move( fn ), MakeThisFunc( __OnUnitChangeBegin /**/ ) ); return this; }
	UiMatrixEditor* UiMatrixEditor::OnUnitChangeEnd   /**/( OnUnitChangeEnd_t /**/ && fn ) { m_OnUnitChangeEnd   /**/ = SetEventCallback<Ui::IMatrixEditor::OnUnitChangeEnd   /**/>( std::move( fn ), MakeThisFunc( __OnUnitChangeEnd   /**/ ) ); return this; }

}
